### PR TITLE
feat: createBaseRouter in TS

### PR DESCRIPTION
**Pull request by golota60**
This one was the most painful to do :P

### CODE REVIEW

File: src/Redirect.tsx [hunk 1]
```diff
@@ -32,4 +38,10 @@ if (__DEV__) {
   (Redirect.prototype as any).isReactComponent = {};
 }
 
-export default Redirect;
+// This actually doesn't extend a React.Component, but we need consumer to think that it does
+class RedirectType extends React.Component<RedirectOptions> {
+  // @ts-ignore
+  constructor(config: RedirectOptions);
+}
```
Comment (golota60) [inline src/Redirect.tsx:45]: I don't know whether there's a better way to do this. The problem is i need consumer to think this component `extends React.Component`

File: src/Redirect.tsx [hunk 2]
```diff
@@ -1,13 +1,19 @@
...
 import RedirectException from './RedirectException';
-import { Match, RedirectOptions } from './typeUtils';
+import { LocationDescriptor, Match, RedirectOptions } from './typeUtils';
+
+class Redirect implements RedirectOptions {
+  path?: string;
+
+  to: string | ((match: Match) => LocationDescriptor);
+
+  status?: number;
```
Comment (golota60) [inline src/Redirect.tsx:12]: I decided to just declare those on the classes. Technically those are functional changes, but pracically they should not change anything. I'm pretty sure the only difference is that `this` has those initialized as `undefined` when creating the instance, which should not change anything

File: src/Redirect.tsx (hunk 1 above)
Comment (jquense) [reply src/Redirect.tsx:45]: I think this is ok but maybe do: 

```tsx
declare class RedirectType extends React.Component<RedirectOptions> {
  constructor(config: RedirectOptions);
}
```

with the declare which should make it just a type?

Comment (golota60) [reply src/Redirect.tsx]: yep, it works :smile:

**Review by jquense — APPROVED** [review]
LGTM

**Comment by jquense** [comment]
Nice work, I was dreading these weird objects :P

**Pull request merged**